pkg/utils/ldap/reconnect.go

```diff
@@ -0,0 +1,317 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package ldap
+
+// LDAP automatic reconnection mechanism, inspired by:
+// https://gist.github.com/emsearcy/cba3295d1a06d4c432ab4f6173b65e4f#file-ldap_snippet-go
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/go-ldap/ldap/v3"
+	"github.com/rs/zerolog"
+)
+
+var (
+	errMaxRetries = errors.New("max retries")
+)
+
+type ldapConnection struct {
+	Conn  *ldap.Conn
+	Error error
+}
+
+// ConnWithReconnect maintains an LDAP Connection that automatically reconnects after network errors
+type ConnWithReconnect struct {
+	conn    chan ldapConnection
+	reset   chan *ldap.Conn
+	retries int
+	logger  *zerolog.Logger
+}
+
+// Config holds the basic configuration of the LDAP Connection
+type Config struct {
+	URI          string
+	BindDN       string
+	BindPassword string
+	TLSConfig    *tls.Config
+}
+
+// NewLDAPWithReconnect Returns a new ConnWithReconnect initialized from config
+func NewLDAPWithReconnect(config Config) *ConnWithReconnect {
+	conn := ConnWithReconnect{
+		conn:    make(chan ldapConnection),
+		reset:   make(chan *ldap.Conn),
+		retries: 1,
+	}
+	logger := zerolog.Nop()
+	conn.logger = &logger
+	go conn.ldapAutoConnect(config)
+	return &conn
+}
+
+// SetLogger sets the logger for the current instance
+func (c *ConnWithReconnect) SetLogger(logger *zerolog.Logger) {
+	c.logger = logger
+}
+
+func (c *ConnWithReconnect) retry(retries int, fn func(c ldap.Client) error) error {
+	conn, err := c.getConnection()
+
+	if err != nil {
+		return err
+	}
+
+	for try := 0; try <= retries; try++ {
+		if try > 0 {
+			c.logger.Debug().Msgf("retrying attempt %d", try)
+			conn, err = c.reconnect(conn)
+			if err != nil {
+				// reconnection failed stop this attempt
+				return err
+			}
+		}
+		if err = fn(conn); err == nil {
+			// function succeed no need to retry
+			return nil
+		}
+		if !ldap.IsErrorWithCode(err, ldap.ErrorNetwork) {
+			// non network error, stop retrying
+			return err
+		}
+	}
+	return errMaxRetries
+}
+
+// Search implements the ldap.Client interface
+func (c *ConnWithReconnect) Search(sr *ldap.SearchRequest) (*ldap.SearchResult, error) {
+	var err error
+	var res *ldap.SearchResult
+
+	retryErr := c.retry(c.retries, func(c ldap.Client) error {
+		res, err = c.Search(sr)
+		return err
+	})
+
+	if retryErr != nil {
+		return nil, ldap.NewError(ldap.ErrorNetwork, errMaxRetries)
+	}
+	return res, err
```

Comment:
Wouldn't `err` be always `nil`? If it is not, it would be returned as `retryErr` and the function would never come here.

Same counts for all methods that use the `retry` method. (For sake of simplicity I won't comment them all...)

Reply:
Good catch!